### ISSUE TITLE
X11 Error - BadAtom

### ISSUE THREAD

**Issue by tyronx** [opentk/opentk#937]
### Description

Crash on startup for one of our users, apparently caused by a call to Mouse.GetState()

```X11 Error encountered: 
  Error: BadAtom (invalid Atom parameter)
  Request:     17 (0)
  Resource ID: 0x0
  Serial:      23
  Hwnd:        <null>
  Control:     <null>  at System.Environment.get_StackTrace () [0x00000] in <ab4a534ed83547df8a5f88035772715b>:0 
  at System.Windows.Forms.XplatUIX11﻿.HandleError (System.IntPtr display, System.Windows.Forms.XErrorEvent& error_event) [0x00000] in <b8705ebc4ff24f9aacb4536fc682dee5>:0 
  at OpenTK.Platform.X11.Functions.XGetAtomName (System.IntPtr , System.IntPtr ) [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.Platform.X11.XI2MouseKeyboard.UpdateDevices () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.Platform.X11.XI2MouseKeyboard..ctor () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.Platform.X11.XI2Input..ctor () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.Platform.X11.X11Factory.get_InputDriver () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.Platform.X11.X11Factory.CreateMouseDriver () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 ﻿
  at OpenTK.Input.Mouse..cctor () [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at _7M1yj4FQBTJEmfWvf7qNxqjmlto._R5QcbrHJytWHAq0n2UFgdxz1uit ﻿() [0x00000] in <4c2e6b6ed1db4ee89afd2df1fc0c2550>:0 
  at _7M1yj4FQBTJEmfWvf7qNxqjmlto._GqJojisQakCd6FNKy6lKacFO8co (System.Object , OpenTK.FrameEventArgs ) [0x00000] in <4c2e6b6ed1db4ee89afd2df1fc0c2550>:0 
  at OpenTK.GameWindow.OnRenderFrame (OpenTK.FrameEventArgs e) [0x00000] in <d2cd3fbe958341e98eff3ee31a301de2>:0 
  at OpenTK.GameWindow.OnRenderFrameInternal (OpenTK.FrameEventArgs e) [0x00000] in <d2cd3fbe958341e98eff3ee31a301d﻿e2>:0 ```

### Repro steps

Unlikely to be reproducible for anyone but [the user that reported](https://www.vintagestory.at/forums/topic/948-game-doesnt-start-when-clientsettingsjson-is-present/) it.  

### Related information 

* Gentoo GNU/Linux
* Native backend
* OpenTK 3.0
* Mono 5.20.1.19

**Comment by varon** [comment]
Would love a patch for this if we can ever figure out what's causing it.

**Comment by UnknownShadow200** [comment]
Running the program through xtrace might help.

**Comment by tyronx** [comment]
Here are the trace files
[traces.zip](https://github.com/opentk/opentk/files/3573577/traces.zip)

**Comment by jvbsl** [comment]
could you debug this line: https://github.com/opentk/opentk/blob/3.0/src/OpenTK/Platform/X11/XI2MouseKeyboard.cs#L376
And tell use what the value of valuator->label is?

**Comment by tyronx** [comment]
How do i go about doing that? Compile OpenTK in debug mode?

**Comment by jvbsl** [comment]
either that, or with Rider(if you have it) you can debug into decompiled external code. But the last debug output is something which you should be able to get without any additional work, by just looking at the debug output window. Because I'm pretty sure that valuator->label is 0x0(in trace it seems to be, if I read it correctly).
So my guess is that there is some "mouse-device" which is special. therefore the device name would be interesting, which the debug output should tell...

**Comment by UnknownShadow200** [comment]
Yeah definitely seems valuator->label being 0x00 is what causes it.

The last request sent to the X11 server in 035149-strace is:
```writev(25, [{iov_base="\21.\2\0\0\0\0\0"```
Which is a GetAtomName request with an atom of 0x00.

---

What's interesting is that you can also find that *exact same request* in 035736-noconfig-strace. 
I'm 99% sure the only reason it doesn't crash in the noconfig case is because that codepath ends up initialising winforms somehow, which caused [mono's X11 error swallowing](https://github.com/mono/mono/blob/69ebdd8b9384d2cdd9fd72458f14fe3a83ade1ef/mcs/class/System.Windows.Forms/System.Windows.Forms/XplatUIX11.cs#L2120) code to get installed. 

Especially since around 150 lines later in the trace:
```
write(1, "X11 Error encountered: \n  Error:"..., 1024) = 1024
write(1, "river () [0x00000] in <d2cd3fbe9"..., 1024) = 1024
write(1, "1a301de2>:0 \n  at OpenTK.GameWin"..., 796) = 796
```

**Comment by varon** [comment]
Any further updates on this one?

**Comment by NogginBops** [comment]
@UnknownShadow200 sorry to revive this so long after your comment.

I'm not terribly familiar with X11, so it's not entirely clear to me what a good solution for this would be.
We could add an if-statement at that point and check if `valuator->label` is `0x0` and print using something as `unknown atom name` or something like that.
But if it's an issue somewhere else in the code then that is not the greatest solution as it's just a bandaid.

Also of note is #1275 which touches this code, though it doesn't touch the line in question, but it might fix the underlying issue?

**Comment by UnknownShadow200** [comment]
I'm not really sure on the best solution either, not really that familiar with X11.
If you check https://www.x.org/releases/current/doc/man/man3/XIQueryDevice.3.xhtml though, it does explicitly state for XIValuatorClass that `If the label field is not None, the value of label is an Atom
describing the axis.` - so it might just be best to print `unknown atom name` when label is `0x0`.

**Comment by NogginBops** [comment]
Thanks for finding the relevant documentation!
Searching upon the info in that documentation I found some code samples that suggest that printing that this valuator didn't have a label is the correct solution here.
It would still be interesting to find out when X11 decides it wants to create a `XIValuatorClassInfo` with no label, but it doesn't seem critical, especially as this is a debug printout.

So I think making an if-statement is the solution to aim for.

This doesn't completely explain why the original issue appears though, but it could be something timing related in @tyronx that is making this crash only some of the time.

**Comment by tyronx** [comment]
I haven't heard of any new reports about this i'm afraid, so I would close this issue altogether.

**Comment by NogginBops** [comment]
I think we can add the check for the label being 0 as the documentation says it might be 0. 
But yeah we can close this issue with that fix and just ignore the "why did it crash for that one user that one time" question.

**Comment by NogginBops** [comment]
Moving this issue to `3.3.4` so that I can release `3.3.3`.
